### ISSUE TITLE
Inconsistent error handling

### ISSUE THREAD

**Issue by NfNitLoop** [nushell/nushell#10624]
### Describe the bug

I find it a bit confusing how/when errors are handled by Nushell. It seems maybe my confusion was due to [a bug][1]?

[1]: https://discord.com/channels/601130461678272522/601130461678272524/1159660978850508870

### How to reproduce

```nushell
> # 1: Define some function that might fail partway through:
> def foo [] { echo start; ^false; echo end }

> # 2: Default behavior is to stop execution. But not very loudly. Maybe "begin" looks like good output to me?
> foo  
begin

> # 3: I can force a more "loud" error with `do -c`:
> do -c { foo } 
begin
Error: nu::shell::external_command

  × External command failed
   ╭─[entry #66:1:1]
 1 │ def foo [] { echo begin; ^false; echo end }
   ·                           ──┬──
   ·                             ╰── External command failed
   ╰────
  help:

> # 4: When used in a pipe, suddenly the behavior changes. No short-circuiting OR loud failure.
> foo | $"result: ($in)"  # no early exit or loud error.
result: end

> # 5: Even if I try to m force the loud failure, pipes seem to override it:
> do -c { foo } | $"result: ($in)"  # ... even if we try to force it.
result: end
```

### Expected behavior

* My ideal behavior would be for `foo` to fail with the "External command failed" error in all cases without me needing to do anything special. Silent failures are a bit of a pet peeve of mine.  But I think changing that behavior is a ticket all by itself, so I'm happy to set that aside for now.  (But if you agree, can we open that ticket? 😊) 

* Probably need some more detailed docs on how errors are *supposed* to be handled so when some hapless newb like me comes across it they can figure it out. 😅  [The Book] has a [Creating your own errors] section, but doesn't mention how errors are handled by default. The [do for core] page documents some error handling flags, but I really can't tell from them what the behavior should be w.r.t. my `foo` function, or what the default behavior is.  For the record, current behavior is: 

```shell
~> do -i { foo }
begin
~> do -s { foo }
begin
~> do -p { foo }
begin
~> do { foo }
begin
~> do -c { foo }
begin
Error: nu::shell::external_command [etc.]
```

* And once the desired behavior is documented, the remaining bug is making things work that way. 

[The Book]: https://www.nushell.sh/book/
[Creating your own errors]: https://www.nushell.sh/book/creating_errors.html
[do for core]: https://www.nushell.sh/commands/docs/do.html

### Configuration

| key                | value                                          |
| ------------------ | ---------------------------------------------- |
| version            | 0.85.0                                         |
| branch             |                                                |
| commit_hash        |                                                |
| build_os           | macos-aarch64                                  |
| build_target       | aarch64-apple-darwin                           |
| rust_version       | rustc 1.72.0 (5680fa18f 2023-08-23) (Homebrew) |
| cargo_version      | cargo 1.72.0 (26bba4830 2023-08-26)            |
| build_time         | 2023-09-19 18:42:47 +00:00                     |
| build_rust_channel | release                                        |
| allocator          | mimalloc                                       |
| features           | dataframe, default, sqlite, trash, which, zip  |
| installed_plugins  |                                                |

### Additional context

❤️ Nushell!

**Comment by sholderbach** [comment]
Can you narrow the title a bit as to what may behave inconsistently?

**Comment by jntrnr** [comment]
Did some investigation into this and wanted to write it out here before we try to come up with a fix.

I put in an early break on non-zero external commands in `eval_block`:

```rust
                        if let Some(v) = v.pop() {
                            stack.add_env_var("LAST_EXIT_CODE".into(), v.clone());
                            if let Ok(exit_code) = v.as_int() {
                                if exit_code > 0 {
                                    input = PipelineData::empty();

                                    break 'outer;
                                }
                            }
                        }
```

This largely seems to work and, allowing nushell to stop running the custom command when the non-zero exit code is hit.

```nushell
def foo [] { echo start; ^false; echo end }; foo | $"result: ($in)"
```
would output "result:"

The problem is we get (at least) a few failed test cases:
```
    commands::redirection::redirect_err
    commands::redirection::redirection_keep_exit_codes
```

If you look at these test cases, an odd thing happens. Both of these test cases:

```
> cat asdfasdfasdf.txt err> a.txt; cat a.txt
```
and
```
> bash test.sh out> out.txt err> err.txt; echo $env.LAST_EXIT_CODE
```
appear to error first and then try to continue as if nothing happened. This would be the equivalent of bash's `;` behaviour. Unfortunately, Nushell acts more like bash's `&&`, and rewriting the above to use a `&&` style in bash would mean bash wouldn't allow either test case to pass either.

In a way, Nushell's `;` is currently a mix of Bash's `&&` and `;`. When it behaves like `&&` or `;` depends on the scenario. This is pretty confusing, to say the least.

We may want to be extra clear that we're using `&&` as a default and then perhaps allow for something more lenient as well that's separate?

**Comment by NfNitLoop** [comment]
> We may want to be extra clear that we're using && as a default and then perhaps allow for something more lenient as well that's separate?

FWIW, bash's `set -e` mode has something like this:

```
# `man bash`:

       set [--abefhkmnptuvxBCHP] [-o option] [arg ...]
              [...]
              -e      Exit immediately if a simple command (see SHELL GRAMMAR
                      above) exits with a non-zero status.  The shell does not
                      exit if the command that fails is part of the command
                      list immediately following a while or until keyword,
                      part of the test in an if statement, part of a && or ⎪⎪
                      list, or if the command's return value is being inverted
                      via !.  A trap on ERR, if set, is executed before the
                      shell exits.
```

My personal preference would be for something similar in Nushell by default:

* "Exit" (`error make`) immediately on non-zero exit codes **unless**:
  * The user explicitly opts in to allow non-zero errors.  (with `do -i/-p` or some new method)
  * (maybe?) The command was run interactively as a single command. (ex; the [git log] case).
  
  [git log]: https://www.jntrnr.com/exit-codes/
  
----

For some context: The main thing that worries me here is that without throw-like error handling by default, nushell functions become a bit of a liability for building larger functionality. If I've got `foo` that calls `bar` and `bar` uses `baz` in some subexpression, I want to be confident that if there's some unhandled error in that call chain that it gets exposed loudly instead of quietly possibly returning the wrong value.

*Consistent* behavior for errors (this bug) will *allow* Nushell users to know how to handle errors in these cases. (assuming the errors don't get swallowed by pipelines or subexpressions as we see above).  But it would still *require* them to do so *everywhere* that they use executables for some functionality.  For example, check the difference between these two cases:

```
> ^cat asdfasdf |split row " "
cat: asdfasdf: No such file or directory
╭───┬──╮
│ 0 │  │
╰───┴──╯
~> open asdfasdf | split row " "
Error: nu::shell::directory_not_found

  × Directory not found
   ╭─[entry #36:1:1]
 1 │ open asdfasdf | split row " "
   ·      ────┬───
   ·          ╰── directory not found
   ╰────
```

Nushell's built-in `open` does the right thing here.  But if I try to use some executable to get output instead, Nushell confidently passes an incorrect value back to me. I'll only notice it's wrong if I notice the stderr output (which may have scrolled off-screen.)  

OK, forgive my rambling. I think I feel strongly enough about throw-by-default error handling that I'll open a separate ticket for it and copy some of these justifications there. 

I'm fine if we keep **this ticket** focused on just making the current error handling well-defined and consistent.  😊

**Comment by jntrnr** [comment]
This `^cat asdfasdf |split row " "` looks like it's working correctly, though it might be a bit surprising.

That's because when `split row` begins its work, we don't have a non-zero exit code. It's pulling on the external that is streaming information to it. Only after the command has completed do we know that there's an error, but by then `split row` has already begun processing its stdout. So you get both the stderr output as well as the output from `split row`.

The difference between that and internal commands is that for the internal commands we have two phases: setup and stream. When `open` is doing its setup phase, we already know the file isn't there so we can give an early error.

**Comment by jntrnr** [comment]
Here's an example from bash showing a similar idea:

```
> cat asfasdf | echo done
done
cat: asfasdf: No such file or directory
```

**Comment by NfNitLoop** [comment]
@jntrnr:

Bash is unfortunately saddled with decades of backward compatibility requirements so added some changes to error handling as opt-ins.

`set -e` will get you something similar to nu's stop-early behavior. 
`set -o pipefail` gets you something similar to how nu handles this:

```
> open asdfasdf | str length
Error: nu::shell::directory_not_found

  × Directory not found
   ╭─[entry #7:1:1]
 1 │ open asdfasdf | str length
   ·      ────┬───
   ·          ╰── directory not found
   ╰────
```

```
bash-3.2$ cat asdfasdf | echo done && echo ok || echo error
done
cat: asdfasdf: No such file or directory
ok
bash-3.2$ set -o pipefail
bash-3.2$ cat asdfasdf | echo done && echo ok || echo error
done
cat: asdfasdf: No such file or directory
error
```

**Comment by jntrnr** [comment]
@NfNitLoop yeah, the pipefail thing is a good example. Just to check in - does what I'm saying about the stages make sense? Even with pipefail, bash isn't able to do the check for the filepath before the stream begins. You can only catch that there was an error after the stream finishes.

Nushell would be the same way with externals. The best we could do is to catch that there was an error that occurred during the stream, but unfortunately not before the stream began.

**Comment by NfNitLoop** [comment]
> Does what I'm saying about the stages make sense? Even with pipefail, bash isn't able to do the check for the filepath before the stream begins. You can only catch that there was an error after the stream finishes.

👍 It makes sense that you can't "throw" based on the exit code until it's done. Doesn't make sense IMO to say the behavior of `cat … | ` is "correct" if nu pipelines can fail in a stream, but executables can't. 

> The best we could do is to catch that there was an error that occurred during the stream

Of course!  And this would be a big improvement!

Bash's pipefail does have one issue which you might want/have to avoid. It doesn't differentiate between failures due to a closed pipe, vs. the process exiting with an exit code, so there's this annoying case:

```
bash-3.2$ set -o pipefail
bash-3.2$ yes | head -5; echo $?
y
y
y
y
y
141
```

**Comment by jokeyrhyme** [comment]
Loosely related to https://github.com/nushell/nushell/issues/10176 in overall goal, at least
